File: 0901-1000/0981-Time Based Key-Value Store/0981-Time Based Key-Value Store.py
import collections
import bisect
class TimeMap:
    
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.table = collections.defaultdict(list) 

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.table[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        keys = self.table.get(key, None)
        if keys is None:
            return ""
        i = bisect.bisect_right(keys, (timestamp, chr(127)))
        return keys[i - 1][1] if i else ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: 0901-1000/0981-Time Based Key-Value Store/0981-Time Based Key-Value Store.cpp
class TimeMap {
public:
    /** Initialize your data structure here. */
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        table[key][timestamp] = value;
    }
    
    string get(string key, int timestamp) {
        if (!table.count(key)) {
            return "";
        }

        auto it = table[key].upper_bound(timestamp);
        if (it == table[key].begin()) {
            return "";
        }
        return (--it)->second;
    }

private:
    unordered_map<string, map<int, string>> table;
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */
